### ISSUE TITLE
Upgrade web containers to PHP 7.1

### ISSUE THREAD

**Issue by tannerjfco** [drud/ddev#330]
## What happened (or feature request):
PHP 7.0.x Active Support will end Dec 3, 2017, with Security Support ending Dec 3, 2018.

## What you expected to happen:
PHP 7.1.x is the current stable release. We should upgrade the web containers to use it.

We will want to review if there are any serious compatibility issues with Drupal 7/8 or WordPress and PHP 7.1 before implementing the upgrade.

## Related source links or issues:

https://secure.php.net/supported-versions.php

**Comment by rickmanelius** [comment]
* FWIW, we're seeing at least one hosting platforms fully support 7.1. See [Platform.sh](https://platform.sh/2016/12/php-71).
* Here are all [Drupal PHP 7.1 Open issues](https://www.drupal.org/project/issues/search?text=&projects=&assigned=&submitted=&project_issue_followers=&status%5B%5D=Open&issue_tags_op=%3D&issue_tags=PHP+7.1). 
* Here are all [Drupal PHP 7 Open Issues](https://www.drupal.org/project/issues/search?text=&projects=&assigned=&submitted=&project_issue_followers=&status%5B%5D=Open&issue_tags_op=%3D&issue_tags=PHP+7)

TL;DR. I think if we're already supporting 7.0, the delta of new issues on 7.1 is small. Let's proceed!

**Comment by tannerjfco** [comment]
Digging around a bit for WordPress, I don't see any indication of major issues running it on PHP 7.1. At that point it becomes more a question of distinct plugins/modules being compatible.

I think it's fine to move to 7.1, but we should probably consider / decide whether we want to keep a 7.0 tag maintained alongside a 7.1 tag, in case there's a benefit re: compatibility providing the older release.

**Comment by rickmanelius** [comment]
Good to hear/see that there are no critical WP issues. And I'm sure there will be Drupal modules and WordPress plugins that will pop up here and there.

What's the LOE on maintaining both versions? Seems reasonable to do both for users that need to toggle for debugging. Just as long as it's not adding unnecessary time/overhead.

**Comment by tannerjfco** [comment]
It's doable, it wouldn't necessarily add significant *effort* to tasks involving update of the web container, but it will be additional time/overhead nonetheless. Any change we make to the web containers, we'll need to ensure updated tags get pushed for 7.0 and 7.1. We previously managed multiple versions for web, but we may have some initial setup work to get this going again. It's possible there's a better way to handle it from last time now that the versions aren't so different.

**Comment by rfay** [comment]
I recommend that we avoid spanning 2 PHP versions unless we're forced to do so by customer demand.

**Comment by rickmanelius** [comment]
Fair. Let's see if people care first before we spend the effort.

**Comment by tannerjfco** [comment]
When we perform this update, I suggest that we do the following:

- Ensure we have php extension parity with pantheon. https://pantheon.io/docs/application-containers/#whats-in-an-application-container Maybe ask some PHP developers if there are any other extensions we should include
- Consider using the official PHP images as base. They provide mechanisms for adding extensions that may be more approachable to an end user if they need to build their own container w/ a new extension. There is also an official PHP image variant that includes apache, so we might also have a low-effort path to providing an additional server option.

**Comment by rickmanelius** [comment]
Figured it might be useful to roll this out along https://github.com/drud/ddev/issues/463 (Drush for 8.4). Also, note https://github.com/drud/ddev/issues/488 (using maria DB).

**Comment by rfay** [comment]
Reports are inconsistent about running php 7.1 with Drupal 7. The reality is that even PHP 7.0 is not fully compatible yet, AFAICT.  [This issue for php 7.0](https://www.drupal.org/node/2656548) reports the open 7.0 issues and references some 7.1 issues which remain open.

**Comment by rfay** [comment]
Note that Drupal 8.4.0 is actually incompatible with php7.0 out of the box, won't run without php 7.1.  https://www.drupal.org/docs/8/system-requirements/php

This means that our drupal-project instructions are broken at this time. 

This is actually a Drupal *bug* introduced though. and they say it will be fixed: https://www.drupal.org/node/2898119#comment-12189765 does fix it.

However, it leaves us in an insecure place, since we have changed our docs to recommend the drupal-project install technique, but we still have php7.0 as our only php7 environment.

**Comment by rfay** [comment]
@rickmanelius gave this a thumbs-up and approved immediate work on it, especially since Drupal 8.4 out of the box currently has a bug making it require php7.1. @tannerjfco accepted it on his plate.

**Comment by rickmanelius** [comment]
And there was much rejoicing!
